### PR TITLE
Unicode Rethink [Do Not Merge]

### PR DESCRIPTION
**Pull request by dabrahams**
Lays foundation for String rethink.

The following failures are known to exist

FAIL: Swift(macosx-x86_64) :: compiler_crashers/28616-swift-parser-parseexprsequence-swift-diag-bool-bool.swift (957 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/UnicodeTrie.swift.gyb (1048 of 9168)
FAIL: ["GraphemeClusterSegmentation/UnicodeSpec"]
FAIL: Swift(macosx-x86_64) :: stdlib/String.swift (1050 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/InputStream.swift.gyb (5805 of 9168)
[     FAIL ] ReadLine.Unicode/strippingNewline=false
[     FAIL ] ReadLine.Unicode/strippingNewline=true
FAIL: ["Unicode/strippingNewline=false", "Unicode/strippingNewline=true"]
FAIL: Swift(macosx-x86_64) :: api-digester/source-stability.swift (5983 of 9168)
FAIL: Swift(macosx-x86_64) :: stdlib/Character.swift (7023 of 9168)
[     FAIL ] Character.RoundTripping/Random
FAIL: ["RoundTripping/Random"]
FAIL: Swift(macosx-x86_64) :: stdlib/Renames.swift (7048 of 9168)

### CODE REVIEW

**Comment by dabrahams** [comment]
@swift-ci Please test Linux platform

File: test/Prototypes/Unicode.swift [hunk 1]
```diff
@@ -49,6 +50,25 @@ extension ParsedUnicode {
 }
 
 //===--- Missing stdlib niceties ------------------------------------------===//
+postfix operator ^
```
Comment (airspeedswift) [inline test/Prototypes/Unicode.swift]: 😐

File: test/Prototypes/string_prototype/StringComparison.swift [hunk 2]
```diff
@@ -58,23 +58,88 @@ where
       // be affected by churn in future Unicode versions.
       //
       // TODO: cost/benefit investigation of more complex check here...
-      return codeUnit < 0xD800
+      return codeUnit < 0xD800 || codeUnit > 0xDFFF
```
Comment (dabrahams) [inline test/Prototypes/string_prototype/StringComparison.swift]: Please describe the more complex check in the comment

**Review by dabrahams — COMMENTED** [review]
Very excited to see this series of commits, @milseman!

File: test/Prototypes/string_prototype/StringComparison.swift (hunk 2 above)
Comment (milseman) [inline test/Prototypes/string_prototype/StringComparison.swift]: I adjusted the surrounding comments in my latest push. Sorry, I've been trying to wrap everything up today before heading out, and the history got a little chaotic at the end.

**Comment by airspeedswift** [comment]
@swift-ci  please test linux platform

File: test/Prototypes/string_prototype/StringCompatibility.swift [hunk 3]
```diff
@@ -62,3 +72,31 @@ extension String {
   }
 }
 
+// TODO: this on the new String
```
Comment (dabrahams) [inline test/Prototypes/string_prototype/StringCompatibility.swift]: Please use `#if false` for these kinds of things to cut down on VCS churn

File: test/Prototypes/string_prototype/Substring.swift.gyb [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    get {
+      return String(self).${property}
+    }
+  }
+% end
+
+  // public typealias ${ViewPrefix}Index = String.Valid${ViewPrefix}View.Index
+
+  public var unicodeScalars: String.UnicodeScalarView {
+    get {
```
Comment (airspeedswift) [inline test/Prototypes/string_prototype/Substring.swift.gyb]: is there a style/future-proofing reason to not elide the `get` here?

File: test/Prototypes/string_prototype/main.swift [hunk 5]
```diff
@@ -335,4 +335,15 @@ testSuite.test("substring") {
   expectEqualSequence("ello world", tail)
 }
 
+import Foundation
+testSuite.test("bridging") {
+  defer { _debugLogging = false }
+  _debugLogging = false
+  let s : main.String
```
Comment (milseman) [inline test/Prototypes/string_prototype/main.swift]: Should that be: `defer { _debugLogging = true }`?

Comment (dabrahams) [inline test/Prototypes/string_prototype/main.swift]: No, the intention was to change line 341 to `= true` to turn on logging for just this case.

File: test/Prototypes/AU3.swift.gyb [hunk 6]
```diff
@@ -814,6 +852,70 @@ extension String : Hashable {
...
+    if let l = content._latin1, l.isASCII() {
+      return l.withExistingUnsafeBuffer(body)!
+    }
+    */
+    // FIXME: add a utf8 view to UnicodeContent so this can be faster in many
+    // cases.
+    var result = ContiguousArray<CChar>()
+    let source = Counted(content.utf8)
+    result.reserveCapacity(numericCast(source.count + 1))
+    result.append(contentsOf: source.lazy.map { numericCast($0) })
```
Comment (airspeedswift) [inline test/Prototypes/AU3.swift.gyb]: Doesn't this need to be `Int8.init(bitPattern:)` rather than `numericCast`, which will trap for UTF8 values > 128?

**Comment by shahmishal** [comment]
@dabrahams What is the status of this PR?

**Comment by dabrahams** [comment]
> On Oct 2, 2018, at 10:44 AM, Mishal Shah <notifications@github.com> wrote:
> 
> @dabrahams <https://github.com/dabrahams> What is the status of this PR?
> 
> 

I'm not doing anything with it; feel free to close if it's causing clutter.

**Pull request closed without merging**